source/_posts/2017-10-09-guarding-an-open-door-authorizing-your-graphql-fields.markdown

```diff
@@ -0,0 +1,348 @@
+def call(memo, visit_type, irep_node)
+  irep_node.definition # nil
+end
+```
+
+The first time we hit this method, our `irep_node` will be the `query` node (one of three [root types](http://graphql.org/learn/execution/#root-fields-resolvers) in our schema: `query`, `mutation`, and `subscription`). The second time, we'll get the `delayedJobs` node:
+
+```ruby
+def call(memo, visit_type, irep_node)
+  irep_node.definition.name # "delayedJobs"
+end
+```
+
+Now that we know which field the user is trying to get access to, we just need to surface the `current_user` object so we can check for the relevant permissions. That's as easy as attaching the `current_user` to the query context:
+
+```ruby
+# Called from the controller endpoint handling GraphQL requests
+Graph::Schema.execute(
+  query: "your_query_string_here",
+  context: { current_user: current_user },
+  operation_name: "your_operation_name_here",
+  variables: { ... }
+)
+```
+
+Then, we can call the `current_user` in our query analyzer:
+
+```ruby
+def call(memo, visit_type, irep_node)
+  irep_node.query.context[:current_user]
+end
+```
+
+Checking for the right permissions goes something like this:
+
+```ruby
+def call(memo, visit_type, irep_node)
+  current_user = irep_node.query.context[:current_user]
+  requested_node_name = irep_node.definition.name.to_sym
+  ability = Ability.new(current_user) # Using CanCan
+  memo[:unauthorized_nodes] ||= []
+  memo[:unauthorized_nodes] << irep_node if ability.cannot?(:access, requested_node_name)
+  memo
+end
+```
+
+Finally, we can return an `AnalysisError` once query analysis has completed:
+
+```ruby
+def final_value(memo)
+  unauthorized_node_names = memo[:unauthorized_nodes].map { |node| node.definition.name }
+  GraphQL::AnalysisError.new("You do not have permission to access #{unauthorized_node_names.join(', ')}")
+end
+```
+
+If we want to get fancy, we can build some [instrumentation](http://graphql-ruby.org/queries/instrumentation.html) to attach metadata to each field indicating which action and subject we'd like to authorize against.
+
+**Key Takeaways:**
+
+- Ahead-of-time authorization is used when the conditions for access do not depend on the values returned by each field
+
+- Authorization happens after the query has been parsed, but before the query is executed. A query analyzer makes a good hook for authorization checks
+
+- Ahead-of-time authorization usually takes two parameters: The query context and the metadata for each field
+
+## Runtime Authorization
+Let's go back to the third-grade class in the lobby. We're not worried if they're playing with the gold bar on display; in fact, it's encouraged (what else is there to do at the Mint?). We'd be a little more concerned if we found a third-grader in the vaults, though.
+
+In that case, we can't forbid everyone from handling our gold. But we can be very clear about *which* ingots can be touched.
+
+Let's take a look at a different query:
+
+```
+query {
+  user(id: 123) {
```

Comment:
Should we use a model that doesn't exist in our db (maybe related to a mint)? Or is user generic enough? Just double-checking.

Reply:
My assumption is that `User` is generic enough that it doesn't reveal too much about our schema, and that it's relevant enough to be a good example, but I'm definitely open to changing these examples.

Reply:
Ok, I'm fine with this example unless someone else pipes up.